### PR TITLE
MAINT: Fixup random bounds checking code

### PR DESCRIPTION
**Pull request by seberg**
This tries to clean up the bounds checking for random integers a bit.  By doing the following things:

1. Add paths to check whether `np.can_cast` is true and casting is safe.  In that case, the values cannot possibly be out of bounds.
2. When they _can_ be out of bounds and we are dealing with int64/uint64, assume overflows are plausible/likely.  In that case always use the slow path by manually going via Python floats `int` (manual casting using `int()`) to ensure correct bounds checking.
3. Since we converted the arrays, we can then safely compare them.  This also means I move the `is_open` into the loop in the int64 path (to match the other path).

Since conversion errors don't match the `ValueError` adds a `ValueError` explicitly on the lower (path is hit in tests).

Doing this lets the test suite pass largely with weak promotion enabled.  For the upcasting paths I need to add `np.int32({{ub}})` to ensure correct casting when weak.

(Test suite still chokes in a few places in random due to `isnan(very_large_integer)` failing no the expected way.)

---

I suspect this might make some outrageous inputs a bit safer, but not sure.  Unfortunately, this is a bit tricky code :(.

### CODE REVIEW

**Comment by mattip** [comment]
It makes sense, maybe a test would help clarify what the use case is?

**Comment by seberg** [comment]
This makes the tests pass with weak promotion, which prevents using object arrays in some paths (and thus things go haywire).

I can try to think of things that fail currently (tomorrow) without the change.  There are relatively good tests probing the bounds checking, but I *suspect* there are some holes that this fixes.

**Comment by charris** [comment]
close/reopen

**Comment by charris** [comment]
Hypothesis is unhappy when running the debug test, valid test values are hard to find.

**Comment by seberg** [comment]
Strange, the test runs very quickly locally, some of the other tests which run in 0.3 seconds locally take 30+ seconds (`test_float_modulus_exact`, it is a slow test locally too, but 100 times faster).  What surprises me though is that the overall test run seems fine.

Restarting the run once (more?), I guess in principle hypothesis might be using this code internally so small(?) changes in performance are plausible.

**Comment by seberg** [comment]
Hmmm, this time it went through.  `numpy/core/tests/test_scalarmath.py::TestModulus::test_float_modulus_exact` seems always absurdly slow on these CI runs.  It seems like a slow test, but I cannot seem reproduce the extreme slowness locally (tried mainly debug-python+pytest-cov) and no particular step in the test seemed to clearly dominate.  I am wondering if it may be related to the Python version.

Doesn't quite help with hypothesis reaching its 1second timeout here sometimes, but I still don't see that it can really be related.

I tried running the specific test locally in main and this branch, and I do not see any performance change though.

File: numpy/random/_bounded_integers.pyx.in [hunk 1]
```diff
@@ -165,50 +172,69 @@ cdef object _rand_{{nptype}}_broadcast(object low, object high, object size,
...
+        flat = low_arr_orig.flat
+        low_data = <{{nptype}}_t *>np.PyArray_DATA(low_arr)
+        cnt = np.PyArray_SIZE(low_arr)
+        for i in range(cnt):
+            lower = int(flat[i])
+            if lower < {{lb}} or lower > {{ub}}:
+                raise ValueError('low is out of bounds for {{nptype}}')
+            low_data[i] = lower
+
+    del low_arr_orig
```
Comment (charris) [inline numpy/random/_bounded_integers.pyx.in:207]: Does this just decrement the reference?

File: numpy/random/_bounded_integers.pyx.in [hunk 2]
```diff
@@ -165,50 +172,69 @@ cdef object _rand_{{nptype}}_broadcast(object low, object high, object size,
...
+    if np.can_cast(low_arr_orig, np.{{otype}}):
+        low_arr = <np.ndarray>np.PyArray_FROM_OTF(low_arr_orig, np.{{npctype}}, np.NPY_ALIGNED)
+    else:
+        low_arr = <np.ndarray>np.empty_like(low_arr_orig, dtype=np.{{otype}})
+        flat = low_arr_orig.flat
+        low_data = <{{nptype}}_t *>np.PyArray_DATA(low_arr)
+        cnt = np.PyArray_SIZE(low_arr)
+        for i in range(cnt):
+            lower = int(flat[i])
+            if lower < {{lb}} or lower > {{ub}}:
```
Comment (charris) [inline numpy/random/_bounded_integers.pyx.in:203]: Does this work for numpy type lb, ub?

File: numpy/random/_bounded_integers.pyx.in [hunk 3]
```diff
@@ -165,50 +172,69 @@ cdef object _rand_{{nptype}}_broadcast(object low, object high, object size,
...
-        highm1_data = <{{nptype}}_t *>np.PyArray_DATA(highm1_arr)
+        high_arr = np.empty_like(high_arr_orig, dtype=np.{{otype}})
+        flat = high_arr_orig.flat
+        high_data = <{{nptype}}_t *>np.PyArray_DATA(high_arr)
         cnt = np.PyArray_SIZE(high_arr)
-        flat = high_arr.flat
         for i in range(cnt):
-            # Subtract 1 since generator produces values on the closed int [off, off+rng]
-            closed_upper = int(flat[i]) - 1
+            closed_upper = int(flat[i]) - is_open
```
Comment (charris) [inline numpy/random/_bounded_integers.pyx.in:217]: Is `is_open` correctly promoted? Hopefully not to float :)

File: numpy/random/_bounded_integers.pyx.in [hunk 4]
```diff
@@ -165,50 +172,69 @@ cdef object _rand_{{nptype}}_broadcast(object low, object high, object size,
...
             if closed_upper > {{ub}}:
                 raise ValueError('high is out of bounds for {{nptype}}')
             if closed_upper < {{lb}}:
                 raise ValueError(format_bounds_error(closed, low_arr))
-            highm1_data[i] = <{{nptype}}_t>closed_upper
+            high_data[i] = closed_upper
 
-    if np.any(np.greater(low_arr, highm1_arr)):
-        raise ValueError(format_bounds_error(closed, low_arr))
+        is_open = 0  # we applied is_open in this path already
```
Comment (charris) [inline numpy/random/_bounded_integers.pyx.in:224]: Why is this needed?

File: numpy/random/_bounded_integers.pyx.in [hunk 5]
```diff
@@ -165,50 +172,69 @@ cdef object _rand_{{nptype}}_broadcast(object low, object high, object size,
...
+    del high_arr_orig
+
+    # Since we have the largest supported integer dtypes, they must be within
+    # range at this point; otherwise conversion would have failed.  Check that
+    # it is never true that `high <= low`` if closed and `high < low` if not
+    if not is_open:
+        low_high_comp = np.greater
+    else:
+        low_high_comp = np.greater_equal
+
```
Comment (charris) [inline numpy/random/_bounded_integers.pyx.in:235]: Everything isn't closed at this point?

**Comment by charris** [comment]
I think this is OK, but not quite sure :)

**Comment by charris** [comment]
Any possibility of a float being cast to int?

File: numpy/random/_bounded_integers.pyx.in (hunk 2 above)
Comment (seberg) [reply numpy/random/_bounded_integers.pyx.in:203]: lb and ub are integer constants and not the numpy type.

File: numpy/random/_bounded_integers.pyx.in (hunk 3 above)
Comment (seberg) [reply numpy/random/_bounded_integers.pyx.in:217]: hehe, its a `bint`, so I think so yes.

File: numpy/random/_bounded_integers.pyx.in (hunk 4 above)
Comment (seberg) [reply numpy/random/_bounded_integers.pyx.in:224]: This path has to assign to int64/uint64, so it must do the `- 1` for is_open and that means we don't have to do it later.  The safe-cast path does it later, though.

File: numpy/random/_bounded_integers.pyx.in (hunk 1 above)
Comment (seberg) [reply numpy/random/_bounded_integers.pyx.in:207]: Yeah, maybe there is no point, happy to delete.

File: numpy/random/_bounded_integers.pyx.in (hunk 5 above)
Comment (seberg) [reply numpy/random/_bounded_integers.pyx.in:235]: Its especially hard to read in the diff, but the safe-cast path doesn't adjust so is not only the unsafe casting point (going via `int()`) does.

**Comment by charris** [comment]
Thanks Sebastian.
